ports/esp32/machine_hw_spi.c

```diff
@@ -392,123 +398,87 @@ STATIC void machine_hw_spi_print(const mp_print_t *print, mp_obj_t self_in, mp_p
+            mp_arg_error_terse_mismatch();
+            #else
+            mp_raise_msg_varg(&mp_type_TypeError, MP_ERROR_TEXT("'%q' argument required"), spi_allowed_args[ARG_id].qst);
+            #endif
+        }
 
-    mp_arg_val_t args[MP_ARRAY_SIZE(allowed_args)];
-    mp_arg_parse_all(n_args, pos_args, kw_args, MP_ARRAY_SIZE(allowed_args),
-        allowed_args, args);
-    int8_t sck, mosi, miso;
+        // Check if id is valid
+        const mp_int_t spi_id = args[ARG_id].u_int;
+        if (1 <= spi_id && spi_id <= MICROPY_HW_SPI_MAX) {
+            default_pins = &machine_hw_spi_default_pins[spi_id - 1];
+        } else {
+            mp_raise_msg_varg(&mp_type_ValueError, MP_ERROR_TEXT("SPI(%d) doesn't exist"), spi_id);
+        }
 
-    if (args[ARG_sck].u_obj == MP_OBJ_NULL) {
-        sck = -2;
-    } else if (args[ARG_sck].u_obj == mp_const_none) {
-        sck = -1;
+        // Replace -1 non-pin args with default values
+        static const mp_int_t defaults[] = { 500000, 0, 0, 8, MICROPY_PY_MACHINE_SPI_MSB };
+        for (int i = ARG_baudrate; i <= ARG_firstbit; i++) {
+            if (args[i].u_int == -1) {
+                args[i].u_int = defaults[i - ARG_baudrate];
+            }
+        }
     } else {
-        sck = machine_pin_get_id(args[ARG_sck].u_obj);
+        // id should not be specified when calling init on existing object
+        if (args[ARG_id].u_int != -1) {
+            #if MICROPY_ERROR_REPORTING <= MICROPY_ERROR_REPORTING_TERSE
+            mp_arg_error_terse_mismatch();
+            #else
+            // TODO better error message
+            mp_raise_TypeError(MP_ERROR_TEXT("extra keyword arguments given"));
+            #endif
+        }
     }
 
-    if (args[ARG_miso].u_obj == MP_OBJ_NULL) {
-        miso = -2;
-    } else if (args[ARG_miso].u_obj == mp_const_none) {
-        miso = -1;
-    } else {
-        miso = machine_pin_get_id(args[ARG_miso].u_obj);
+    // Replace pin args with default/current values for new vs init call, respectively
+    for (int i = ARG_sck; i <= ARG_miso; i++) {
+        if (args[i].u_obj == MP_OBJ_NULL) {
+            args[i].u_int = new ? default_pins->array[i - ARG_sck] : -2;
+        } else if (args[i].u_obj == mp_const_none) {
+            args[i].u_int = -1;
+        } else {
+            args[i].u_int = machine_pin_get_id(args[i].u_obj);
+        }
     }
+}
 
-    if (args[ARG_mosi].u_obj == MP_OBJ_NULL) {
-        mosi = -2;
-    } else if (args[ARG_mosi].u_obj == mp_const_none) {
-        mosi = -1;
-    } else {
-        mosi = machine_pin_get_id(args[ARG_mosi].u_obj);
-    }
+STATIC void machine_hw_spi_init(mp_obj_base_t *self_in, size_t n_args, const mp_obj_t *pos_args, mp_map_t *kw_args) {
+    machine_hw_spi_obj_t *self = (machine_hw_spi_obj_t *)self_in;
+
+    mp_arg_val_t args[MP_ARRAY_SIZE(spi_allowed_args)];
+    mp_arg_parse_all(n_args, pos_args, kw_args, MP_ARRAY_SIZE(spi_allowed_args),
```

Comment:
It might work (and would be simpler) to disallow the `id` argument here, by doing:
```py
mp_arg_parse_all(n_args, pos_args, kw_args, MP_ARRAY_SIZE(spi_allowed_args) - 1, spi_allowed_args + 1, args + 1);
```
That is, skip the `id` argument when passing the `spi_allowed_args` array.

Then, `machine_hw_spi_argcheck()` won't need to check `args[ARG_id].u_int` in this case.

Reply:
Yes, that should work.  And I can avoid the need to put error messages already part of mp_arg_parse_all in the code here too.  I bet that saves a bunch of code.

Reply:
And done.  Saved a bunch of code.  I hated copying the error messages out of the library argument parsing code.

After this, there was this block of `if(new) { }` and the common function, and there was really no point to moving that block out of make_new, so it put it back, which saved some more code.

All this only saved 19 more bytes that the previous version.  This is because the refactored arg checking code was small enough gcc decided to inline it into both call sites.  It be a bit more without that.

Reply:
And done. Saved a bunch of code. I hated copying the error messages out of the library argument parsing code.

After this, there was this block of if(new) { } and the common function, and there was really no point to moving that block out of make_new, so it put it back, which saved some more code.

All this only saved 19 more bytes that the previous version. This is because the refactored arg checking code was small enough gcc decided to inline it into both call sites. It be a bit more without that.